File: dev-c/0012-integer-to-roman.c
size_t _write(size_t write_index, char* output, char c1, char c5, char c10, short n) {
    switch(n) {
        case 0: 
            break;

        case 1:
            output[write_index ++] = c1;
            break;

        case 2:
            output[write_index ++] = c1;
            output[write_index ++] = c1;
            break;

        case 3:
            output[write_index ++] = c1;
            output[write_index ++] = c1;
            output[write_index ++] = c1;
            break;

        case 4:
            output[write_index ++] = c1;
            output[write_index ++] = c5;
            break;

        case 5:
            output[write_index ++] = c5;
            break;

        case 6:
            output[write_index ++] = c5;
            output[write_index ++] = c1;
            break;

        case 7:
            output[write_index ++] = c5;
            output[write_index ++] = c1;
            output[write_index ++] = c1;
            break;

        case 8:
            output[write_index ++] = c5;
            output[write_index ++] = c1;
            output[write_index ++] = c1;
            output[write_index ++] = c1;
            break;

        case 9:
            output[write_index ++] = c1;
            output[write_index ++] = c10;
            break;

    }

    return write_index;
}

char* intToRoman(const int num) {
    char* result = (char*) malloc(16 * sizeof(char));
    size_t write_index = 0;

    write_index = _write(write_index, result, 'M', ' ', ' ', (short) (num / 1000)       );
    write_index = _write(write_index, result, 'C', 'D', 'M', (short) (num % 1000) / 100 );
    write_index = _write(write_index, result, 'X', 'L', 'C', (short) (num % 100)  / 10  );
    write_index = _write(write_index, result, 'I', 'V', 'X', (short) (num % 10)         );

    result[write_index] = '\0';
    return result;
}

File: dev-rust/0012-integer-to-roman.rs
impl Solution {

    fn encode_digit(c1: char, c5: char, c10: char, n: i32) -> String {
        return match n {
            0 => "".to_string(),
            1 => String::from_iter([c1]),
            2 => String::from_iter([c1, c1]),
            3 => String::from_iter([c1, c1, c1]),
            4 => String::from_iter([c1, c5]),
            5 => String::from_iter([c5]),
            6 => String::from_iter([c5, c1]),
            7 => String::from_iter([c5, c1, c1]),
            8 => String::from_iter([c5, c1, c1, c1]),
            9 => String::from_iter([c1, c10]),
            _ => "".to_string()
        }
    }

    pub fn int_to_roman(num: i32) -> String {
        let mut result: String = String::new();

        result += &Self::encode_digit('M', ' ', ' ', (num / 1000)       );
        result += &Self::encode_digit('C', 'D', 'M', (num % 1000) / 100 );
        result += &Self::encode_digit('X', 'L', 'C', (num % 100)  / 10  );
        result += &Self::encode_digit('I', 'V', 'X', (num % 10)         );   
      
        return result;
    }
}

